util/osdiag/osdiag_windows.go

```diff
@@ -328,3 +340,144 @@ func getModuleInfo() (map[string]moduleInfo, error) {
+	}
+	if ret != 0 {
+		return 0, errUnexpectedResult
+	}
+
+	return result, nil
+}
+
+type wsaProtocolInfoOutput struct {
+	Description     string   `json:"description,omitempty"`
+	Version         int32    `json:"version"`
+	AddressFamily   int32    `json:"addressFamily"`
+	SocketType      int32    `json:"socketType"`
+	Protocol        int32    `json:"protocol"`
+	ServiceFlags1   string   `json:"serviceFlags1"`
+	ProviderFlags   string   `json:"providerFlags"`
+	Path            string   `json:"path,omitempty"`
+	PathErr         error    `json:"pathErr,omitempty"`
+	Category        string   `json:"category,omitempty"`
+	CategoryErr     error    `json:"categoryErr,omitempty"`
+	BaseProviderID  string   `json:"baseProviderID,omitempty"`
+	LayerProviderID string   `json:"layerProviderID,omitempty"`
+	Chain           []uint32 `json:"chain,omitempty"`
+}
+
+func getWinsockLSPInfo() (map[uint32]wsaProtocolInfoOutput, error) {
+	protocols, err := enumWinsockProtocols()
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[uint32]wsaProtocolInfoOutput, len(protocols))
+	for _, p := range protocols {
+		v := wsaProtocolInfoOutput{
+			Description:   windows.UTF16ToString(p.ProtocolName[:]),
+			Version:       p.Version,
+			AddressFamily: p.AddressFamily,
+			SocketType:    p.SocketType,
+			Protocol:      p.Protocol,
+			ServiceFlags1: fmt.Sprintf("0x%08X", p.ServiceFlags1), // Serializing as hex string to make the flags easier to decode by human inspection
+			ProviderFlags: fmt.Sprintf("0x%08X", p.ProviderFlags),
+		}
+
+		switch p.ProtocolChain.ChainLen {
+		case windows.BASE_PROTOCOL:
+			v.BaseProviderID = p.ProviderId.String()
+		case windows.LAYERED_PROTOCOL:
+			v.LayerProviderID = p.ProviderId.String()
+		default:
+			v.Chain = p.ProtocolChain.ChainEntries[:p.ProtocolChain.ChainLen]
+		}
+
+		// Queries that are only valid for base and layered protocols (not chains)
+		if v.Chain == nil {
+			path, err := p.path()
+			if err == nil {
+				v.Path = strings.ToLower(path)
+			} else {
+				v.PathErr = err
+			}
+
+			category, err := p.categoryFlags()
+			if err == nil {
+				v.Category = fmt.Sprintf("0x%08X", category)
+			} else if !errors.Is(err, windows.WSAEINVALIDPROVIDER) {
+				// WSAEINVALIDPROVIDER == "no category info found", so we only log
+				// errors other than that one.
+				v.CategoryErr = err
+			}
+		}
+
+		// Chains reference other providers using catalog entry IDs, so we use that
+		// value as the key in our map.
+		result[p.CatalogEntryId] = v
+	}
+
+	return result, nil
+}
+
+func enumWinsockProtocols() ([]wsaProtocolInfo, error) {
+	// Get the required size
+	var errno int32
+	var bytesReqd uint32
+	ret := wscEnumProtocols(nil, nil, &bytesReqd, &errno)
+	if ret != _SOCKET_ERROR {
+		return nil, errUnexpectedResult
+	}
+	if e := windows.Errno(errno); e != windows.WSAENOBUFS {
+		return nil, e
+	}
+
+	// Allocate
+	szEntry := uint32(unsafe.Sizeof(wsaProtocolInfo{}))
+	buf := make([]wsaProtocolInfo, bytesReqd/szEntry)
```

Comment:
Should there be a check that `bytesReqd` is a multiple of `szEntry`?

Reply:
We actually deal with that by computing `bufLen` independently of  `bytesReqd` in the following line to account for any potential rounding down during the integer division.